src/toil/cwl/cwltoil.py

```diff
@@ -682,83 +907,259 @@ def toil_make_tool(
+    index: Dict[str, str],
+    existing: Dict[str, str],
+    file_store_id: str
 ) -> str:
-    """Get path to input file from Toil jobstore."""
-    if not file_store_id.startswith("toilfs:"):
-        return file_store.jobStore.getPublicUrl(
-            file_store.jobStore.importFile(file_store_id, symlink=True)
+    """
+    Set up the given file or directory from the Toil jobstore at a file URI
+    where it can be accessed locally.
+
+    Run as part of the ToilCommandLineTool setup, inside jobs on the workers.
+
+    :param index: Maps from downloaded file path back to input Toil URI.
+
+    :param existing: Maps from file_store_id URI to downloaded file path.
+    """
+
+    if file_store_id.startswith("toildir:"):
+        # This is a file in a directory, or maybe a directory itself.
+        # See ToilFsAccess and upload_directory.
+        # We will go look for the actual file in the encoded directory
+        # structure which will tell us where the toilfs: name for the file is.
+
+        parts = file_store_id[len("toildir:"):].split('/')
+        contents = json.loads(base64.urlsafe_b64decode(parts[0].encode('utf-8')).decode('utf-8'))
+
+        for component in parts[1:]:
+            if component != '.':
+                # Index into the contents
+                contents = contents[component]
+
+        if isinstance(contents, str):
+            # This is a reference to a file, so go get it.
+            return toil_get_file(file_store, index, existing, contents)
+        else:
+            # We actually need to fetch the whole directory to a path somewhere.
+            dest_path = file_store.getLocalTempDir()
+            # Populate the directory
+            download_structure(file_store, index, existing, contents, dest_path)
+            # Return where we put it, but as a file:// URI
+            return schema_salad.ref_resolver.file_uri(dest_path)
+    elif file_store_id.startswith("toilfs:"):
+        # This is a plain file with no context.
+        src_path = file_store.readGlobalFile(FileID.unpack(file_store_id[len("toilfs:"):]), symlink=True)
+        # TODO: shouldn't we be using these as a cache?
+        index[src_path] = file_store_id
+        existing[file_store_id] = src_path
+        return schema_salad.ref_resolver.file_uri(src_path)
+    elif file_store_id.startswith("_:"):
+        # Someone is asking us for an empty temp directory.
+        dest_path = file_store.getLocalTempDir()
+        return schema_salad.ref_resolver.file_uri(dest_path)
+    else:
+        raise RuntimeError(
+            f'Cannot obtain file {file_store_id} from host '
+            f'{socket.gethostname()}; all imports must happen on the '
+            f'leader!'
         )
-    src_path = file_store.readGlobalFile(FileID.unpack(file_store_id[7:]), symlink=True)
-    index[src_path] = file_store_id
-    existing[file_store_id] = src_path
-    return schema_salad.ref_resolver.file_uri(src_path)
+
+def download_structure(
```

Comment:
It could, but I don't think I'll have time to write it before I leave. You might have to do it, or I can do it when I get back. This one won't be as simple because we have to make up a file store and find it a place to download to. We might actually want to redesign this a bit if we really intend to test it.